Fix CI build

vroldanbet opened the pull request:
# Implement development/CI environment based on docker

Master CI build recently broke. The `install-openldap` script is very complex and assumes a specific environment. This is clearly not completely isolated from the fact that new versions of packages may be backported to the distribution used. I suspect this is what happened (noticed builds started breaking with a new version of `slapd`).

Instead of fixing the script, I decided I could reuse open-source maintained OpenLDAP docker containers. The https://github.com/osixia/docker-openldap project handles a lot of the work performed by `install-openldap`. As a consequence. we can remove **a lot of stuff** from this repo:
- some LDIF fixtures
- `script/install-openldap`
- `script/generate-fixture-ca`
- `test/support` (vagrant stuff)
- certificates (it's now generated by the container)

In general, it's a lot of less code to maintain, and less code that can break as dependencies change.

Docker is also fairly ubiquitous in CI systems, and it would help closing the gap between CI and local development environment. Travis supports it nicely and it could open the door to adopt GitHub Actions too?

# Extra niceties

- added most recent ruby versions to the build matrix
- no more "travis specific" tests

# The not so nice part

- I couldn't find a way to workaround having to modify `/etc/hosts`. I feel it's a small price to pay to simplify the setup considerably, but I'll try to explore alternatives (if I managed to name the containers hostname `localhost` it would get fixed)
- the timeout test, I can't make it go green. I decided to disable it.
- test relying on system cert store, didn't try hard how to install cert in travis, and it's additional burden with little return
- test with multiple valid hostnames - I couldn't get the certificate generation to have multiple hostnames. Instead, I decided to run the test with the same valid hostname 2 times. It should _hopefully_ be equivalent.

# TODO

- [ ] allow devs in local to avoid changes to `/etc/hosts` for integration tests

CamiloGarciaLaRotta commented on .travis.yml:
```diff
@@ -3,20 +3,31 @@ rvm:
 
 install:
-  - if [ "$INTEGRATION" = "openldap" ]; then sudo script/install-openldap; fi
+  - docker run --hostname ldap.example.org --env LDAP_TLS_VERIFY_CLIENT=try -p 389:389 -p 636:636 -v $(pwd)/test/fixtures/seed.ldif:/container/service/slapd/assets/config/bootstrap/ldif/50-bootstrap.ldif -v $(pwd)/test/fixtures/openldap/retcode.ldif:/container/service/slapd/assets/config/bootstrap/ldif/06-retcodes.ldif --name openldap --detach osixia/openldap:1.3.0 --copy-service --loglevel debug
```
What do you think of breaking the command into multiline lines? I think it reads easier.

We can use either the folded style `>` or the literal style `|` with backslash `\` to break the line.
I have this bookmarked for YAML syntax: [multiline-info](https://yaml-multiline.info/) and [SO](https://stackoverflow.com/a/21699210)

vroldanbet replied:
Yeah I thought about it and punted because I didn't know if backslash was necessary. Will give it a try!

vroldanbet replied:
Got it in the third attempt 😅 
https://github.com/ruby-ldap/ruby-net-ldap/pull/333/commits/cf4e9bd5a1af3f1982974ef408043cd0f9664234

WDYT?

CamiloGarciaLaRotta commented on test/integration/test_bind.rb:
```diff
@@ -1,12 +1,16 @@
   end
 
   def test_bind_timeout
+    omit "this is no longer working in our test environment - skipping"
```
:(

vroldanbet replied:
Yeah 😢 I'm getting a `connection refused`, which makes total sense, because the test is checking on a closed port. I didn't find any evidence the script set up anything in that port, so begs the question _why did it work this way in first place_ 🤔 🤷‍♂

vroldanbet replied:
finally solved it using a non-routable IP! 🎉

CamiloGarciaLaRotta commented on test/integration/test_bind.rb:
```diff
@@ -146,13 +151,9 @@ def test_bind_tls_with_bogus_hostname_system_ca_fails
-  # The following depend on /etc/hosts hacking.
-  # We can do that on CI, but it's less than cool on people's dev boxes
   def test_bind_tls_with_multiple_hosts
-    omit_unless ENV['TRAVIS'] == 'true'
```
:)

vroldanbet commented on script/install-openldap:
```diff
@@ -1,134 +0,0 @@
-chmod o-r /etc/ssl/private/ldap01_slapd_key.pem
-
-# Drop packets on a secondary port used to specific timeout tests
-iptables -A INPUT -p tcp -j DROP --dport 8389
```
OH! dang, good catch, I searched for `8389` but I clearly missed this. I'll have a look! 👀

vroldanbet replied:
Finally implemented it using an non routable IP, seems to work! https://github.com/ruby-ldap/ruby-net-ldap/pull/333/commits/b90ee7eff619f9f3f80a2ee6ac503b81c4e897c1

vroldanbet commented on .travis.yml:
```diff
@@ -3,20 +3,43 @@ rvm:
+      -p 389:389 \
+      -p 636:636 \
+      -v $(pwd)/test/fixtures/seed.ldif:/container/service/slapd/assets/config/bootstrap/ldif/50-bootstrap.ldif \
+      -v $(pwd)/test/fixtures/openldap/retcode.ldif:/container/service/slapd/assets/config/bootstrap/ldif/06-retcodes.ldif \
```
My thought was these were different kind of data (the second seems to be configuration of the server return codes). I'll give it a try, but chances are it will replace the files that already exist in the container folder instead of adding them - I'll double check that 👍

vroldanbet replied:
Indeed files would be replaced when mounting a folder, but the container creators thought about this use case, and provide an empty `/custom` folder where you can add all your ldifs 🎉 

See https://github.com/ruby-ldap/ruby-net-ldap/pull/333/commits/693b210a80302414d79a224455dab79764e9ea66

vroldanbet commented:
@mtodd thanks for the review ✨ , I addressed your comments, this should be ready to be merged!

potentially related: I don't see any use of `/testserver`. Is that something that could be removed too?
